src/scala/org/pantsbuild/scoverage/report/ScoverageReport.scala

```diff
@@ -0,0 +1,254 @@
+          coverage.apply(measurements)
+          Some(coverage)
+
+        case false =>
+          logger.error("Coverage file did not exist")
+          None
+
+      }
+    } else {
+      logger.error("Data dir did not exist!")
+      None
+    }
+
+  }
+
+  /**
+   * Writes coverage reports usign the specified source path to the specified report directory.
+   */
+  private def writeReports(coverage: Coverage, options: ReportOptions): Unit = {
+    val sourceDir = new File(options.sourcePath)
+    val reportDir = new File(options.reportPath)
+    val reportDirHtml = new File(options.reportPath + "/html")
+    val reportDirXml = new File(options.reportPath + "/xml")
+
+    if (sourceDir.exists) {
+      if (options.cleanOld && reportDir.exists) {
+        logger.info(s"Nuking old report directory [$reportDir].")
+        FileUtils.deleteDirectory(reportDir)
+      }
+
+      if (!reportDir.exists) {
+        logger.info(s"Creating HTML report directory [$reportDirHtml]")
+        reportDirHtml.mkdirs
+        logger.info(s"Creating XML report directory [$reportDirXml]")
+        reportDirXml.mkdirs
+      }
+
+      if (options.writeHtml) {
+        logger.info(s"Writing HTML scoverage reports to [$reportDirHtml]")
+        new ScoverageHtmlWriter(Seq(sourceDir), reportDirHtml, None).write(coverage)
+      }
+
+      if (options.writeXml) {
+        logger.info(s"Writing XML scoverage reports to [$reportDirXml]")
+        new ScoverageXmlWriter(Seq(sourceDir), reportDirXml, false).write(coverage)
+        if (options.writeXmlDebug) {
+          new ScoverageXmlWriter(Seq(sourceDir), reportDirXml, true).write(coverage)
+        }
+      }
+    } else {
+      logger.error(s"Source dir [$sourceDir] does not exist")
+    }
+
+    logger.success(s"Statement coverage: ${coverage.statementCoverageFormatted}%")
+    logger.success(s"Branch coverage:    ${coverage.branchCoverageFormatted}%")
+  }
+
+  def main(args: Array[String]): Unit = {
+    Settings.parser1.parse(args, Settings()) match {
+      case Some(settings) =>
+        val reportOptions = ReportOptions(settings)
+
+        reportOptions.loadDataDir match {
+          case false =>
+            loadAggregatedCoverage(reportOptions.measurementsPath, reportOptions) match {
+              case Some(cov) =>
+                logger.success("Coverage loaded successfully.")
+                if (reportOptions.writeReports) {
+                  writeReports(cov, reportOptions)
+                }
+
+              case None => logger.error("Failed to load coverage.")
+            }
+          case true =>
+            loadCoverage(reportOptions.dataPath) match {
+              case Some(cov) =>
+                logger.success("Coverage loaded successfully!")
+                if (reportOptions.writeReports) {
+                  writeReports(cov, reportOptions)
+                }
+              case _ => logger.error("Failed to load coverage")
```

Comment:
All of the failure cases in this matchshould cause the process to exit unsuccessfully.

To exit unsuccessfully, the process should either:
1. raise an exception
2. call `System.exit(1)`

I think that rather than having the `loadCoverage` method return an `Option`, you might consider having it just directly raise an exception. But if you don't feel comfortable with that and would prefer a functional style, a better thing to return would probably be `Either[Coverage, String]` (where the right hand side would be an error message to exit with). https://www.scala-lang.org/api/2.12.8/scala/util/Either.html

Reply:
`loadCoverage` now raises an exception